src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,982 @@
+        ldout(m_cct, 5) << "lost exclusive lock, retrying remote" << dendl;
+        continue;
+      }
+    }
+    break;
+  }
+  if (r < 0) {
+    lderr(m_cct) << "migrare failed: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = unset_migration(dst_image_ctx);
+  if (r < 0) {
+    return r;
+  }
+
+  r = enable_mirroring();
+  if (r < 0) {
+    return r;
+  }
+
+  r = remove_src_image();
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::list_snaps() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = snap_list(m_src_image_ctx, m_snaps);
+  if (r < 0) {
+    lderr(m_cct) << "failed listing snapshots: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  for (auto &snap : m_snaps) {
+    bool is_protected;
+    r = snap_is_protected(m_src_image_ctx, snap.name.c_str(), &is_protected);
+    if (r < 0) {
+      lderr(m_cct) << "failed retrieving snapshot status: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+    }
+    if (is_protected) {
+      lderr(m_cct) << "image has protected snapshot '" << snap.name << "'"
+                   << dendl;
+      return -EBUSY;
+    }
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::set_migration() {
+  m_src_image_ctx->ignore_migrating = true;
+
+  if (m_src_old_format) {
+    return v1_set_migration();
+  } else {
+    return v2_set_migration();
+  }
+}
+
+template <typename I>
+int Migrate<I>::v1_set_migration() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = cls_client::migration_set(&m_src_io_ctx, m_src_header_oid,
+                                    m_src_migration_spec);
+  if (r < 0) {
+    lderr(m_cct) << "failed to set migration header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  static_assert(sizeof(RBD_HEADER_TEXT) == sizeof(RBD_MIGRATE_HEADER_TEXT),
+                "length of rbd banners must be the same");
+
+  bufferlist header;
+  header.append(RBD_MIGRATE_HEADER_TEXT);
+
+  r = m_src_io_ctx.write(m_src_header_oid, header, header.length(), 0);
+  if (r < 0) {
+    lderr(m_cct) << "failed writing " << m_src_header_oid
+                 << " banner: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = tmap_rm(m_src_io_ctx, m_src_image_name);
+  if (r < 0) {
+    lderr(m_cct) << "failed removing " << m_src_image_name << " from tmap: "
+                 << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::v2_set_migration() {
+  ldout(m_cct, 10) << dendl;
+
+  v2_unlink_image();
+
+  int r = cls_client::migration_set(&m_src_io_ctx, m_src_header_oid,
+                                    m_src_migration_spec);
+  if (r < 0) {
+    lderr(m_cct) << "failed to set migration header: " << cpp_strerror(r)
+                 << dendl;
+    if (r == -EOPNOTSUPP) {
+      v2_relink_image();
+    }
+    return r;
+  }
+
+  r = m_src_image_ctx->operations->update_features(RBD_FEATURE_MIGRATING, true);
```

Comment:
I thought about this. The only thing why I decided to do this by `operations->update_features()` was that it would automatically call `ImageCtx::notify_update()` so other clients would be informed about the migrating feature bit was set.

But taking into account all other related things you mentioned below, it really looks like better to do in cls_rbd.cc and then call `notify_update()` manually.